rcl_yaml_param_parser/src/parser.c

```diff
@@ -696,74 +167,113 @@ void rcl_yaml_node_struct_fini(
   allocator.deallocate(params_st, allocator.state);
 }
 
-
 ///
-/// Find node entry index in parameters' structure
+/// TODO (anup.pemmaiah): Support Mutiple yaml files
+///
 ///
-static rcutils_ret_t find_node(
-  const char * node_name,
-  rcl_params_t * param_st,
-  size_t * node_idx)
+/// Parse the YAML file and populate params_st
+///
+bool rcl_parse_yaml_file(
+  const char * file_path,
+  rcl_params_t * params_st)
 {
-  assert(NULL != node_name);
-  assert(NULL != param_st);
-  assert(NULL != node_idx);
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    file_path, "YAML file path is NULL", return false);
 
-  for (*node_idx = 0U; *node_idx < param_st->num_nodes; (*node_idx)++) {
-    if (0 == strcmp(param_st->node_names[*node_idx], node_name)) {
-      // Node found.
-      return RCUTILS_RET_OK;
-    }
+  if (NULL == params_st) {
+    RCUTILS_SAFE_FWRITE_TO_STDERR("Pass an initialized parameter structure");
+    return false;
+  }
+
+  yaml_parser_t parser;
+  int success = yaml_parser_initialize(&parser);
+  if (0 == success) {
+    RCUTILS_SET_ERROR_MSG("Could not initialize the parser");
+    return false;
+  }
+
+  FILE * yaml_file = fopen(file_path, "r");
+  if (NULL == yaml_file) {
+    yaml_parser_delete(&parser);
+    RCUTILS_SET_ERROR_MSG("Error opening YAML file");
+    return false;
+  }
+
+  yaml_parser_set_input_file(&parser, yaml_file);
+
+  namespace_tracker_t ns_tracker;
+  memset(&ns_tracker, 0, sizeof(namespace_tracker_t));
+  rcutils_ret_t ret = parse_file_events(&parser, &ns_tracker, params_st);
+
+  fclose(yaml_file);
+
+  yaml_parser_delete(&parser);
+
+  rcutils_allocator_t allocator = params_st->allocator;
+  if (NULL != ns_tracker.node_ns) {
+    allocator.deallocate(ns_tracker.node_ns, allocator.state);
   }
-  // Node not found, add it.
-  rcutils_allocator_t allocator = param_st->allocator;
-  param_st->node_names[*node_idx] = rcutils_strdup(node_name, allocator);
-  if (NULL == param_st->node_names[*node_idx]) {
-    return RCUTILS_RET_BAD_ALLOC;
+  if (NULL != ns_tracker.parameter_ns) {
+    allocator.deallocate(ns_tracker.parameter_ns, allocator.state);
   }
-  rcutils_ret_t ret = node_params_init(&(param_st->params[*node_idx]), allocator);
+
   if (RCUTILS_RET_OK != ret) {
-    allocator.deallocate(param_st->node_names[*node_idx], allocator.state);
-    return ret;
+    rcl_yaml_node_struct_fini(params_st);
+    return false;
```

Comment:
@brawner one thing I noticed while testing something else is that this function is finalizing its output argument on error, invalidating the output argument without the caller knowing. I'll push a fix.

Reply:
See https://github.com/ros2/rcl/pull/776.